### ISSUE TITLE
Release 2.0

### ISSUE THREAD

**Issue by samoht** [mirage/irmin#658]
# Irmin release task list (last updated 04.10)

### Irmin

- [x] Wait for ocaml-git release and integrate it with #868 
- [x] Complete the CHANGELOG
- [x] Complete the API migration guide
- [x] Add upper bounds for revdeps in opam-repository
- [x] Solve the close issue
- [x] Fix the tests

### irmin.io

- [x] Remove irmin-js and irmin-go from irmin.io tutorial
- [ ] Write a blog post on index, irmin-pack
- [x] Write a blog post on irmin 2.0

### Index

- [x] ~~Remove the `hash_size`~~ (delayed)
- [x] Fix readonly `v` when store does not exist and test it
- [x] Review the documentation
- [x] Have non-blocking merges

## Release process

- [x] Release `index.0.1.0`
- [x] Release `irmin-*.2.0.0~rc1`
- [x] Update https://irmin.org domain name
- [x] Publish the documentation
- [x] Publish the blog post about irmin 2.0
- [ ] Publish the blog post about irmin-pack
- [x] Publish the blog post about GraphQL

**Comment by avsm** [comment]
Looks good. If you have an opam remote that tracks the packages, I can also refresh docs.mirage.io to include bleeding edge versions of the Irmin 2.0 RCs.

**Comment by samoht** [comment]
After discussing with @zshipko it appears than we don't want to ship the OCaml graphQL client right now. They still need a few more tweaks before being totally happy with it. So it's better to not ship something that we know will change for sure in the very near future.

**Comment by hannesm** [comment]
@samoht thanks! about the MirageOS compatibility:
- @andreas fixed graphql-parser already :D
- I think https://github.com/mirage/irmin/pull/659 is what we discussed in Marrakesh which should work (NB: I haven't tested this, `Sys.getcwd ()` returns an absolute path, now `.` is a relative one -- is that ok?)
- the third issue is that Canopy doesn't hold their store anymore (you should be able to reproduce by testing https://github.com/Engil/Canopy/pull/95 and see whether it comes up with a webserver, for me it fails to read `.config/uuid` from the git remote (which is there according to printfs after the clone)):
```OCaml
let store_config = Irmin_mem.config ()
let repo _ = Store.Repo.v store_config

let store () =
  match remote_branch () with
  | None        -> repo () >>= Store.master
  | Some branch -> repo () >>= fun r -> Store.of_branch r branch
```

This `store ()` leads to an empty repository most of the time.

**Comment by zshipko** [comment]
Based on the outstanding PRs and issues it seems do-able to plan the release for **April 3rd**. This gives us two weeks to get everything together, which should be enough for what's listed below.

If anyone thinks that timeline is too aggressive then there's no harm in extending it, however I think it would be best to make the release as soon as possible. Also, let me know if there's anything I'm missing here.

## Timeline

- **March 29th**
    * finish outstanding PRs
    * finalize `CHANGES.md` and migration guide
    * make sure tutorial is completely up-to-date
- **April 1st/2nd**
    * final code review
- **April 3rd**
    * publish Irmin 2.0.0 to opam

After this we can start on the 2.0.x release, which will culminate with a blog-post/larger announcement.

## Release related materials

- [CHANGES.md](https://github.com/zshipko/irmin/blob/changes-2.0/CHANGES.md) (needs editing/review)
- [Irmin 2.0 migrations guide](https://gist.github.com/zshipko/88b4ef34815624ca7dbffd809e8eeec7) (needs editing/review)
- [Tutorials](https://github.com/zshipko/irmin-tutorial) (needs final review)


## Outstanding issues and PRs

- `ocaml-git` issue when pushing to git_daemon ([ocaml-git#342](https://github.com/mirage/ocaml-git/issues/342))
- Small changes to support MirageOS ([#659](https://github.com/mirage/irmin/pull/659))
- Small update to `irmin-graphql` server to clean-up API (in progress: [#660](https://github.com/mirage/irmin/pull/660), @zshipko)

**Comment by samoht** [comment]
The timeline looks good -- we even be a bit more aggressive and plan for a 2.0 next week if everything goes well, and start to check that all the constraints of the rev-deps are in place.

The only big unknown for me is the issue that @hannesm is seeing while using Canopy. Would be great to check if this can be reproduced and understand the cause/impact a bit more.

**Comment by zshipko** [comment]
@hannesm I'm not able to reproduce this using Engil/Canopy#95 - is there something else I need to do other than run `./canopy`?

**Comment by hannesm** [comment]
@zshipko I tested this on solo5-hvt, not sure whether the unix backend will take another path in dependencies. I will re-try this weekend in a fresh switch and record the steps. Thanks for attempting to reproduce.

**Comment by zshipko** [comment]
Okay, I am able to reproduce the issue using the solo5-hvt backend - I will look into this a bit more

**Comment by hannesm** [comment]
@zshipko good to know, did you make progress on that?

**Comment by samoht** [comment]
@dinosaure told me that it should be fixed by https://github.com/mirage/ocaml-git/pull/338 (to check)

**Comment by hannesm** [comment]
@samoht I'm not sure. the clone/checkout/pull works fine -- and on Unix, even Canopy with your PR seems to work fine... but I've not read through git/irmin changes to have an expert opinion on this..

**Comment by samoht** [comment]
@dinosaure @zshipko did you get more understanding of the root cause of the MirageOS issue? Should we delay the release? (it's perfectly fine to delay the release, but I just would like to understand what happens :p)

**Comment by zshipko** [comment]
Unfortunately, I haven't been able to find out anything else about the Canopy/MirageOS issue, so I'm not sure about the release. I'm hoping @dinosaure has more insight!

**Comment by dinosaure** [comment]
I'm currently on some issues of `ocaml-git`. At this stage, I'm currently on the weird bug about `hvt`, in other side, I'm waiting a feedback from @hannesm about a deep change on the smart protocol where I mostly fix it blindfolded. If you can post-pone the release about one day, I can give you a report tomorrow.

**Comment by samoht** [comment]
That sounds good, let's delay the release for a few days to be sure we haven't missed anything :-)

**Comment by hannesm** [comment]
for MirageOS, it would be nice to add support of a `kv_ro` and `kv_rw` to the mirage command line utility which uses irmin+remote git as storage (see [`generic_kv_rw/ro`](https://docs.mirage.io/mirage/Mirage/index.html#val-generic_kv_ro)). at configure-time I'd like to select which kv implementation is used (unix/mem/git), and runtime provide the necessary data (file path, maximum size, remote url) as boot parameters. NB: @linse and me tried and failed, and atm [hardcoding irmin-kv/rw into unikernels](https://github.com/roburio/caldav/tree/5e7b504f90ce0c8ce1d44c02b72738c5bae97d29/mirage) :/ (see https://github.com/hannesm/functoria-playground for some failed attempts at dealing with functoria)

**Comment by samoht** [comment]
@hannesm agreed but I guess this can be done after the 2.0 release.

**Comment by hannesm** [comment]
@samoht it can be delayed, yes.

in respect to irmin-mirage and git-mirage and the new release, I'd be really happy if `push and pull` would work on kvm/hvt/freestanding, as well as `Canopy`.

## Canopy

Applying [your patch](https://github.com/Engil/Canopy/pull/95) does clone the store, but doesn't keep the data in the `store ()` below (i.e. when `pull` is called (by main), and i dump the values after the ``Sync.pull_exn t upstream `Set``, i see data -- but subsequent operations that use `store ()` get an empty store -- canopy then complains on startup that it can't find `.config/uuid`)).
```
let store_config = Irmin_mem.config ()
let repo _ = Store.Repo.v store_config

let store () =
  match remote_branch () with
  | None        -> repo () >>= Store.master
  | Some branch -> repo () >>= fun r -> Store.of_branch r branch
```

## [push and pull](https://gist.github.com/hannesm/50e3c60d1e95d04aba14b7d1f3f64297)

this is a unikernel that uses the mirage-kv API and irmin+git in the backend. It listens on tcp port 1234 and 1235.
- if a client connects on port 1234, it pulls (Store.connect .. KV.fold) the given remote and prints it to the client.
- if a client connects on port 1235 and enters a "key`:`value", this key -> value is pushed to the remote repository

issues i saw today:
- pull works via https and the git protocol
- push doesn't work

good news:
- git push works with the git protocol (tested locally against a git_daemon 2.19.1) if i use https://github.com/mirage/ocaml-git/pull/350 (which needs some more released packages ([encore](https://github.com/mirage/encore/pull/15), ke (I think)) and changes here in #668 (needed by git master)) (in addition @dinosaure checkseum should  have a release with the ["cross-compilation" fixes](https://github.com/mirage/checkseum/compare/v0.0.3...master) which are already merged to master)
- git push via https worked back in january on this repository https://github.com/roburio/testcalendar ... now i get:
```
2019-04-03 20:58:24 -00:00: DBG [irmin.git] push https://miragebot:foobar23@github.com/miragebot/testdata
2019-04-03 20:58:24 -00:00: DBG [git.sync.http] Launch the GET request to https://github.com/miragebot/testdata/info/refs?service=git-receive-pack.
2019-04-03 20:58:24 -00:00: DBG [cohttp] Send a request to https://github.com/miragebot/testdata/info/refs?service=git-receive-pack.
2019-04-03 20:58:24 -00:00: ERR [git.sync.http] The HTTP decoder returns an error: (`No_assert_predicate #predicate).
2019-04-03 20:58:24 -00:00: ERR [application] error while pushing data (`Smart (`No_assert_predicate #predicate))
```

**Comment by dinosaure** [comment]
@hannesm mirage/ocaml-git#351 should fix one of your reported error. The last case when you want to push via HTTPS on `testcalendar` is not fixed yet however. However, could you test PR?

**Comment by hannesm** [comment]
to report back, I think that most issues are fixed with https://github.com/mirage/ocaml-git/pull/351

I switched to using `Irmin.Store.KV` and `Sync` directly (instead of KV_RW) to have tight control over push and pull operations -- since I wanted to support mutliple writers. What I'd like:
- a pretty printer for the error result of Sync.push (see https://github.com/roburio/unikernels/blob/32767a5af8a6b281b1e9c6a38cde399ee0a452b1/primary-git/unikernel.ml#L102-L107)
- a pretty printer for the error result of Sync.pull (see https://github.com/roburio/unikernels/blob/32767a5af8a6b281b1e9c6a38cde399ee0a452b1/primary-git/unikernel.ml#L19-L22)
- some documentation what these steps:
```OCaml
    let config = Irmin_mem.config () in
    Store.Repo.v config >|= Store.master
```

are actually doing -- i.e. does the `Store.Repo.v` and/or `Store.master` already do network communication (they can't since they don't know the remote yet, or?)? why are they in the Lwt.t monad?

**Comment by samoht** [comment]
@hannesm does this mean that the current KV_RW interface is useless? Or is there anything that we can do to give finer control on push/pull?

Good catch for the pretty-printers, I'll add these tomorrow.

@dinosaure do you understand what happens with git/solo5?

**Comment by hannesm** [comment]
@samoht the current KV_RW interface is fine if you're the only writer - there's explicit `push` control (`batch` / `write`), but no control over `pull` at all.

the `pull` could either be periodically (not so nice) or a `KV_RW` should offer a specific `sync` operation, called by the application once it receives a webhook / notification that the backend changed. there's also need to control merge in the multiple writer store, but only during `set` and `remove` (these operations atm offer only a write_error, but no way to merge that).

**Comment by zshipko** [comment]
After being away for the last week I'm trying to understand what is needed to make this release happen.

Is https://github.com/mirage/ocaml-git/issues/353 what needs attention right now? I am happy to spend some time trying to understand what's happening there and can also take care of the pretty-printers and documentation update requested by @hannesm [above](https://github.com/mirage/irmin/issues/658#issuecomment-480896908).

It looks like this means we will need to postpone the `irmin` 2.0.0 until after the next `ocaml-git` update since these latest changes depend on the development version of `ocaml-git` (see: https://github.com/mirage/irmin/pull/668). This is unfortunate, but if Canopy, and other stable applications are broken then it makes no sense to move forward until these things are dealt with.

**Comment by hannesm** [comment]
so, is there any release plan? if nobody apart from myself is interested in irmin + git remote on MirageOS, then  you should go ahead and release without supporting that. I'm rather low on time atm (including the foreseeable future), and thus won't be able to debug and test much further (I hope the above provided examples are helpful). What occured to me is that `irmin-mirage` now depends on graphql, and I'm not sure why this should be the case -- maybe the irmin-mirage opam package should be split into multiple ones?

**Comment by yomimono** [comment]
I'm interested in irmin + git remote on MirageOS but similarly not ready to take advantage of it in the near future, so I shouldn't block a release.

**Comment by zshipko** [comment]
Well we're waiting on a new `ocaml-git` release and @samoht is out of town for the next two weeks, so I plan on spending this time debugging any known issues/preparing and updating documentation.

At this point it's best to make a new issue for any problems you encounter (rather than adding it to this thread) - that way we can keep track of what needs attention between the 2.0.0 and 2.0.x releases.

I don't think we're really thinking about releasing a broken version of `irmin-git` + remote on MirageOS, however since I'm not specifically using it for anything it's hard to find and reproduce bugs. One things that could help with this is some kind of unikernel test-suite (I will need to think about this!)

As for the `graphql` dependency in `irmin-mirage` - this was added to both the `unix` and `mirage` backends in hopes to eventually replace the current HTTP server. I suppose that `Irmin_mirage.Graphql` could live in a package like `irmin-mirage-graphql` or something, is this an issue because of executable size or something else?

**Comment by hannesm** [comment]
@zshipko sounds good to me. yes, the graphql is not only a build-dependency (which adds to build time), but also runtime code size for functionality i do not plan to use. I like to have systems without any http (server nor client) code in them ;)

unikernel test suite: well, i linked to smaller ones i developed just for irmin, there's also some work on an e2e mirage test system (see https://github.com/mato/e2e-mirage-solo5/), which may be useful to have with irmin unikernels (where you may need a git daemon elsewhere as well).

I commented since I saw that the release plan was early April, now it's rather end of April. I have some irmin+git unikernels deployed, thanks to @dinosaure there are fewer issues in git now, and fewer pins needed, but i guess e.g. the canopy issue should be looked into more deeply (you [tried canopy on unix](https://github.com/mirage/irmin/issues/658#issuecomment-475424237) and it worked, without unix it didn't -- this would worry me if I'd be in the process of releasing a library which behaviour should be the same independent of the platform).

**Comment by samoht** [comment]
What's the status of the release? @zshipko what are the last blockers?

**Comment by zshipko** [comment]
@samoht At this point we're mainly waiting on the `ocaml-git` release before we can start our release. 

Other than that, there is https://github.com/mirage/irmin/issues/671 and https://github.com/mirage/irmin/issues/681, which also seem important to address.

**Comment by dinosaure** [comment]
So `encore` is under the pipe to release, `eqaf` too and issues about `ocaml-git` are fixed. So no big blockers, I think on my side. It's just I take a time to push on the release button.

**Comment by samoht** [comment]
no pb, good to know there is no blockers :-) Let us know if you need some help to release things.

**Comment by hannesm** [comment]
i guess the canopy issue (see https://github.com/mirage/irmin/issues/658#issuecomment-474888703 https://github.com/mirage/irmin/issues/658#issuecomment-475448609) still needs debugging and fixing, no?

(see https://github.com/Engil/Canopy/pull/95 for the specific PR to canopy which according to @zshipko works on mirage+unix, but fails with mirage+solo5)

**Comment by zshipko** [comment]
Yeah, sorry @hannesm! That is definitely another issue that should be looked at before the release. I haven't had any further luck in tracking down the bug though, so another set of eyes could be helpful here.

**Comment by samoht** [comment]
@dinosaure any news regarding the ocaml-git release?

**Comment by dinosaure** [comment]
Just the ownership bug on KVM with Canopy and then, I think it's done.

**Comment by hannesm** [comment]
> Just the ownership bug on KVM with Canopy

what's `ownership` and on `KVM`? if you elaborate, I'm willing to help debugging (if necessary and possible).

**Comment by dinosaure** [comment]
@hannesm it seems that with `ocaml-git` and `solo5`, a `Cstruct.t` was shared by two `lwt` threads - where, in `ocaml-git`, we start with a pool of threads and each has a bunch of buffers. But I did not reproduce it nor try to reproduce it.

**Comment by dinosaure** [comment]
Ok, the bug does exists anymore on `ocaml-git` (don't ask me why), so we should not have any others blockers for this release.

**Comment by hannesm** [comment]
I can confirm that Canopy works (with @dinosaure irmin2 branch) and a current irmin (cloning via https FWIW) -- I'm atm on irmin c930a971828a0560ecf330fa842fb54f59f1f820

so to me, this sounds good to be released! :)

**Comment by samoht** [comment]
That's great!!

@zshipko is everything else ready for the release (changelog, blog post, tutorials, etc)? I'd like to add a few more performance improvements before the release (and to stabilise a bit more `irmin-pack`), but these are not blockers for releasing the rest of the packages.

It might also make sense to integrate https://github.com/zshipko/irmin-tutorial into this repostory (for instance in a `doc/` directory) so that we could use gh-pages to host them.

**Comment by zshipko** [comment]
I think the changelog is up to date, but I will make sure of that tomorrow. Everything else seems ready to go!

`irmin-tutorial` should definitely move into the main repo at some point. My only hesitation right now is the use of `mdBook` which is a big Rust application. It may be possible to replace that with `omd` or something more convenient.

**Comment by hannesm** [comment]
FWIW, when re-trying a moderately modern irmin and git on my Canopy blog, the initial clone works fine, but subsequent pulls/clones lead to errors -- (using this https://github.com/hannesm/hannes.nqsb.io repository -- both via git protocol and via https protocol) -- other git repositories are working nicely, I wonder whether there is still something wrong with pulling from a remote that has some signed commits in it. I unfortunately won't have time to further debug this in the soon future. this may as well be related to https://github.com/mirage/ocaml-git/issues/364 (and actually be an issue in the git implementation rather than irmin).

**Comment by samoht** [comment]
(that git wire protocol is so hard to debug --- would be nice to have traces that every implementation could use to check correctness)

**Comment by Tails** [comment]
What is now still blocking the 2.0 release?

**Comment by avsm** [comment]
@Tails a lot of bugfixes resulting from the Tezos live deployments of the beta releases.  Do you have a particular usecase or deadline that we should we aware of?

**Comment by Tails** [comment]
> @Tails a lot of bugfixes resulting from the Tezos live deployments of the beta releases. Do you have a particular usecase or deadline that we should we aware of?

No I'm just an enthusiast waiting for release, sorry to bother! :)

**Comment by pascutto** [comment]
I updated the task list for the process, we should be able to release within the next couple of weeks !

**Comment by dinosaure** [comment]
`ocaml-git.2.1.1` was released according issues related. Feel free to ship Irmin 2.0.0.

**Comment by yomimono** [comment]
I would like to build and demonstrate an application using Irmin in the waning days of the month. Is there a chance Irmin 2.0 will be released by Thursday 31st October?

**Comment by samoht** [comment]
@yomimono sorry to hear about that. 

The release process is blocked on https://github.com/mirage/index/pull/118 which aims to be merged this week. Once this is done, we will be feature-complete and we could start the release cycle (at least tagging everything and submitting packages to opam).

**Comment by yomimono** [comment]
Great, thanks for the update @samoht!

**Comment by hannesm** [comment]
the above mentioned PR to index is merged now, so does this mean there will be an irmin release promptly? I'd very much appreciate this to be able to remove irmin pins in various projects. (I suspect a git release + #883 would be nice to have in the release, otherwise you'll need to add some upper bounds.)

**Comment by samoht** [comment]
Current status:

- index 1.0 is almost released: https://github.com/ocaml/opam-repository/pull/15283
- irmin 2.0 is almost ready: https://github.com/mirage/irmin/pull/887
- @avsm and @samoht need to add ALIAS and CNAME records to redirect irmin.netlify.io to irmin.io
- @avsm and myself are preparing the irmin v2 post
- @pascutto is preparing a blog post on index and irmin-pack
- @andreas has written a blog post on irmin-graphql and is waiting on the release to publish it

**Comment by avsm** [comment]
Alias records added for irmin.io, and index merged to opam-repo

**Comment by samoht** [comment]
Current status:
- https://irmin.org is live @icristescu [is writing](https://github.com/mirage/irmin.io/pull/27/files) an architecture doc that can be added asynchronously (e.g. it doesn't block the release)
- we are waiting for the [ocaml-git](https://github.com/ocaml/opam-repository/pull/15292) release to be merged -- this is currently blocked on a broken CI due to omp errors (which should be fixed once the CI picks up https://github.com/ocaml/opam-repository/pull/15300)
- I've merged a few last-minutes low risk features (GraphQL API improvement which have been reviewed, irmin-pack support for the CLI, mirage 3.7.1 support, etc)
- blog posts are being written and should be ready on Monday

So given it's already quite late on Friday and CI is a bit broken, let's delay the code+blog post release on Monday.

**Comment by hannesm** [comment]
on your website, you link to uDNS which has now been migrated to mirage/ocaml-dns (and does not per se use irmin) -- instead, there are unikernels in (a) roburio/caldav (b) roburio/unikernels (esp. primary-git) that use irmin (+git) -- maybe worth linking there directly (I also plan to rebase / push to master the respective branches once irmin 2.0 hits opam-repository)!

**Comment by samoht** [comment]
Quick update: we found a weird bug while stress-testing irmin-pack, we are investigating to understand the causes.

**Comment by samoht** [comment]
The root cause of the weird issue has been found and [fixed](https://github.com/mirage/index/pull/132), so I'm starting the release process!

**Comment by samoht** [comment]
PR opened on opam-repository: https://github.com/ocaml/opam-repository/pull/15341

**Comment by samoht** [comment]
```

**Comment by hannesm** [comment]
I think this can be closed now, and thanks again for the release :D
